aws/rust-runtime/aws-config/src/ecs.rs

```diff
@@ -0,0 +1,670 @@
+    use std::time::{Duration, UNIX_EPOCH};
+    use tower::Service;
+
+    fn provider(env: Env, connector: DynConnector) -> EcsCredentialsProvider {
+        let provider_config = ProviderConfig::empty()
+            .with_env(env)
+            .with_connector(connector);
+        Builder::default().configure(&provider_config).build()
+    }
+
+    #[derive(Deserialize)]
+    struct EcsUriTest {
+        env: HashMap<String, String>,
+        result: GenericTestResult<String>,
+    }
+
+    impl EcsUriTest {
+        async fn check(&self) {
+            let env = Env::from(self.env.clone());
+            let uri = Provider::uri(env, Some(BoxCloneService::new(TestDns::default())))
+                .await
+                .map(|uri| uri.to_string());
+            self.result.assert_matches(uri);
+        }
+    }
+
+    #[tokio::test]
+    async fn run_config_tests() -> Result<(), Box<dyn Error>> {
+        let test_cases = std::fs::read_to_string("test-data/ecs-tests.json")?;
+        #[derive(Deserialize)]
+        struct TestCases {
+            tests: Vec<EcsUriTest>,
+        }
+
+        let test_cases: TestCases = serde_json::from_str(&test_cases)?;
+        let test_cases = test_cases.tests;
+        for test in test_cases {
+            test.check().await
+        }
+        Ok(())
+    }
+
+    #[test]
+    fn validate_uri_https() {
+        // over HTTPs, any URI is fine
+        let never = NeverService::new();
+        let mut dns = BoxCloneService::new(never);
+        assert_eq!(
+            validate_full_uri("https://amazon.com", &mut dns)
+                .now_or_never()
+                .unwrap()
+                .expect("valid"),
+            Uri::from_static("https://amazon.com")
+        );
+        // over HTTP, it will try to lookup
+        assert!(
+            validate_full_uri("http://amazon.com", &mut dns)
+                .now_or_never()
+                .is_none(),
+            "DNS lookup should occur, but it will never return"
+        );
+    }
+
+    #[test]
+    fn valid_uri_loopback() {
+        let never = NeverService::new();
+        let mut dns = BoxCloneService::new(never);
+        assert_eq!(
+            validate_full_uri("http://127.0.0.1:8080/get-credentials", &mut dns)
+                .now_or_never()
+                .unwrap()
+                .expect("valid uri"),
+            Uri::from_static("http://127.0.0.1:8080/get-credentials")
+        );
+
+        let err = validate_full_uri("http://192.168.10.120/creds", &mut dns)
+            .now_or_never()
+            .unwrap()
+            .expect_err("not a loopback");
+        assert!(matches!(err, InvalidFullUriError::NotLoopback));
+    }
+
+    #[test]
+    fn all_addrs_local() {
+        let svc = TestDns::with_fallback(vec![
+            "127.0.0.1".parse().unwrap(),
+            "127.0.0.2".parse().unwrap(),
+        ]);
+        let mut svc = BoxCloneService::new(svc);
+        let resp = validate_full_uri("http://localhost:8888", &mut svc)
```

Comment:
No change needed, I'm just curious: In the previous test, we used port 8080 and this one uses 8888. Is that difference significant or just random?